### PR TITLE
WIP: Gather geographical coordinates from pixel coordinates

### PR DESCRIPTION
**Pull request by Xierumeng**
DO NOT MERGE.

Function is untested, correctness unverified.

Unit tests must to be written for the function.

### CODE REVIEW

File: modules/geolocation/geolocation.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+        # Find corresponding geographical coordinate for every valid pixel
+        for i in range(0, np.size(self.__referencePixels)):
+
+            # Not enough pixels to create the map, abort
+            if (validPixelCount < minimumPixelCount):
+                return np.empty(shape=(0, 2, 2))
+
+            # Convert pixel to vector in world space
+            pixel = self.__referencePixels[i]
```
Comment (Shrinjay) [inline modules/geolocation/geolocation.py:48]: Why not just directly access reference pixels? Not major but its a thing

File: modules/geolocation/geolocation.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+            # Scaling in the u, v direction
+            m = 2 * pixel[0] / self.__cameraResolution[0] - 1
```
Comment (Shrinjay) [inline modules/geolocation/geolocation.py:50]: Name these scale_factor_m or something, verbosity helps others understand this.

File: modules/geolocation/geolocation.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+            n = 2 * pixel[1] / self.__cameraResolution[1] - 1
+
+            pixelInWorldSpace3a = c + m * self.__cameraOrientation3u + n * self.__cameraOrientation3v
```
Comment (Shrinjay) [inline modules/geolocation/geolocation.py:53]: Where's c coming from? Would also create a better variable name like with scale_factor above.

File: modules/geolocation/unittests.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+print(pair2)
+
+pairs = np.empty(shape=(0, 2, 2))
+print(pairs)
+pairs = np.concatenate((pairs, [pair1]))
+print(pairs)
+pairs = np.concatenate((pairs, [pair2]))
+print(pairs)
+pairs = np.concatenate((pairs, [pair3]))
+print(pairs)
```
Comment (Shrinjay) [inline modules/geolocation/unittests.py:34]: Let's talk about test cases on Thursday

File: modules/geolocation/geolocation.py [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+        self.__pixelsToGeographicals = np.array([[[0, 0], [0, 0]],
+                                                 [[1, 1], [1, 1]],
+                                                 [[2, 2], [2, 2]],
+                                                 [[3, 3], [3, 3]]])
+
+        return
+
+    # TODO Placeholder, add functionality once we figure out how to convert raw plane data
+    def convert_input(self):
```
Comment (Shrinjay) [inline modules/geolocation/geolocation.py:27]: Does this function call all functions required to convert camera points to geopoints? If not, then we need a function like that which is called by the parent after instantiation. Like the builder pattern, if you're familiar with that.

File: modules/geolocation/geolocation.py [hunk 6]
```diff
@@ -0,0 +1,68 @@
+import numpy as np
+
+class Geolocation:
+    """Locates the geographical position of a set of pixels"""
+
+    # TODO Class members
+    def __init__(self):
```
Comment (Shrinjay) [inline modules/geolocation/geolocation.py:7]: Create a constructor param for the array of points to convert to geocordinates, set pixelsToGeographicals (with new name) to that

File: modules/geolocation/geolocation.py [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+        self.__cameraDirection3c = np.array([0.0, 0.0, -1.0])
+        self.__cameraOrientation3u = np.array([1.0, 0.0, 0.0])
+        self.__cameraOrientation3v = 1 * np.cross(self.__cameraDirection3c, self.__cameraOrientation3u)
+        self.__cameraResolution = np.array([1000, 1000])  # TODO Make global?
+        self.__referencePixels = np.array([[0, 0],
+                                           [0, 1000],
+                                           [1000, 0],
+                                           [1000, 1000]])
+
+        self.__pixelsToGeographicals = np.array([[[0, 0], [0, 0]],
```
Comment (Shrinjay) [inline modules/geolocation/geolocation.py:19]: Are these the pixels we want to convert? If so, please give this a shorter name, I suggest pixelsToConvert

**Review by Shrinjay — CHANGES_REQUESTED** [review]
Codebase looks good, well done, very impressed by your work as always. Made some nitpicky notes on code format and class shape. We'll talk about unit tests on Thursday.

File: modules/geolocation/geolocation.py (hunk 1 above)
Comment (Xierumeng) [reply modules/geolocation/geolocation.py:48]: Makes it clear that we're doing things on a single pixel. Using an alias or pointer was the intention here.

File: modules/geolocation/geolocation.py (hunk 6 above)
Comment (Xierumeng) [reply modules/geolocation/geolocation.py:7]: I thought we were re-using the class to save having to constantly do constructor and destructor calls? Also, pixelsToGeographicals will be populated by the output of the gather_point_pairs() function (currently a placeholder).

File: modules/geolocation/geolocation.py (hunk 7 above)
Comment (Xierumeng) [reply modules/geolocation/geolocation.py:19]: Output of gather_point_pairs() (currently has placeholder values).

File: modules/geolocation/geolocation.py (hunk 5 above)
Comment (Xierumeng) [reply modules/geolocation/geolocation.py:27]: This function converts raw plane data into something the module can do work on (see [documentation](https://docs.google.com/document/d/1VKYIrmWJYfpLPCAjQPiH1J5fnQPtNcjEC2NE5Eg4QxU/edit?usp=sharing)).

Comment (Xierumeng) [reply modules/geolocation/geolocation.py:27]: I haven't yet thought of the name of the function that will run the entire thing from start to finish.

File: modules/geolocation/geolocation.py (hunk 3 above)
Comment (Xierumeng) [reply modules/geolocation/geolocation.py:53]: Mistake, should be replaced with "self.__cameraDirection3c".

**Pull request closed without merging**